contracts/staking/fuse/FuseStakingV4.sol

```diff
@@ -170,16 +189,45 @@ contract FuseStakingV4 is Initializable, OwnableUpgradeable, DSMath {
 			"validator not in approved list"
 		);
 
+		require(_giveBackRatio >= minGivebackRatio, "giveback should be higher or equal to minimum");
 		bool staked = stakeNextValidator(_amount, _validator);
-		stakers[_to] += _amount;
+		updateStakersGivebacks(_to, _amount, _giveBackRatio);
+
 		return staked;
 	}
 
-	function stake(address _validator) public payable returns (bool) {
-		require(msg.value > 0, "stake must be > 0");
-		return _stake(msg.sender, _validator, msg.value);
+	function updateStakersGivebacks(address _to, uint256 _amount, uint256 _giveBackRatio) internal{		
+		if(stakersGivebacks[_to].totalStaked > 0){
+			stakersGivebacks[_to].avgRatio = 
+				weightedAverage(stakersGivebacks[_to].avgRatio, stakersGivebacks[_to].totalStaked, _giveBackRatio, _amount);
+			stakersGivebacks[_to].totalStaked += _amount;
+		}
+		else {
+			stakersGivebacks[_to].totalStaked = _amount;
+			stakersGivebacks[_to].avgRatio = _giveBackRatio;
+		}
+
+		globalGiveback.avgRatio = 
+			weightedAverage(globalGiveback.avgRatio, globalGiveback.totalStaked, _giveBackRatio, _amount);
+		globalGiveback.totalStaked += _amount;
 	}
 
+	/**
+     * @dev Calculates the weighted average of two values based on their weights.
+     * @param valueA The amount for value A
+     * @param weightA The weight to use for value A
+     * @param valueB The amount for value B
+     * @param weightB The weight to use for value B
+     */
+    function weightedAverage(
```

Comment:
Do we have our own math or UTIL lib to add it there? I see there's DSMath but it doesn't use SafeMathUpgradeable